libgadget/uvbg.c

```diff
@@ -0,0 +1,1204 @@
+        uvbg_params.ReionNionPhotPerBary = param_get_double(ps, "ReionNionPhotPerBary");
+        //uvbg_params.AlphaUV = param_get_double(ps, "AlphaUV");
+        uvbg_params.EscapeFraction = param_get_double(ps, "EscapeFraction");
+    }
+    MPI_Bcast(&uvbg_params, sizeof(struct UVBGParams), MPI_BYTE, 0, MPI_COMM_WORLD);
+
+}
+
+double integrand_time_to_present(double a, void *dummy)
+{
+    return 1 / a / hubble_function(&All.CP,a);
+}
+
+//time_to_present in Myr
+double time_to_present(double a)
+{
+#define WORKSIZE 1000
+    gsl_function F;
+    gsl_integration_workspace* workspace;
+    double time;
+    double result;
+    double abserr;
+
+    double hubble;
+    hubble = All.CP.Hubble / All.UnitTime_in_Megayears * All.CP.HubbleParam;
+    //jdavies(second to Myr conversion)
+
+    workspace = gsl_integration_workspace_alloc(WORKSIZE);
+    F.function = &integrand_time_to_present;
+
+    gsl_integration_qag(&F, a, 1.0, 1.0 / hubble,
+        1.0e-8, WORKSIZE, GSL_INTEG_GAUSS21, workspace, &result, &abserr);
+
+    //convert to Myr and multiply by h
+    time = result / (hubble/All.CP.Hubble);
+
+    gsl_integration_workspace_free(workspace);
+
+    // return time to present as a function of redshift
+    return time;
+}
+
+static void assign_slabs()
+{
+    message(0, "Assigning slabs to MPI cores...\n");
+
+    int uvbg_dim = All.UVBGdim;
+    // Allocations made in this function are free'd in `free_reionization_grids`.
+
+    //TODO(jdavies): probably move the first half of this to an initialize function
+    //TODO: do I need to reinitialise this after petapm? I don't see pfft_cleanup anywhere
+    pfftf_init();
+    pfftf_plan_with_nthreads(omp_get_max_threads());
+
+    //TODO: have the flags stored somewhere so it's not both here and in create_plans
+    unsigned pfft_flags = PFFT_PADDED_R2C|PFFT_TRANSPOSED_NONE;
+
+    // Assign the slab size
+    int n_ranks;
+    MPI_Comm_size(MPI_COMM_WORLD, &n_ranks);
+
+    //TODO(jdavies): check if pfft wants in or out dimensions here
+    ptrdiff_t n[3] = {uvbg_dim,uvbg_dim,uvbg_dim};
+    int np[2];
+
+    /* try to find a square 2d decomposition */
+    //TODO(jdavies): ask Simeon about what this is/does and whether I need it
+    int i;
+    for(i = sqrt(n_ranks) + 1; i >= 0; i --) {
+        if(n_ranks % i == 0) break;
+    }
+    np[0] = i;
+    np[1] = n_ranks / i;
+
+    message(0, "Using 2D Task mesh for UVBG %td x %td \n", np[0], np[1]);
+    if( pfftf_create_procmesh_2d(MPI_COMM_WORLD, np[0], np[1], &UVBGgrids.comm_cart_2d) ){
+        endrun(0, "Error: This test file only works with %td processes.\n", np[0]*np[1]);
+    }
+
+    MPI_Comm comm_cart_2d = UVBGgrids.comm_cart_2d;
+
+    // find out what slab each rank should get
+    ptrdiff_t local_ni[3], local_i_start[3];
+    ptrdiff_t local_no[3], local_o_start[3];
+    ptrdiff_t local_n_complex = pfftf_local_size_dft_r2c_3d(n, comm_cart_2d, pfft_flags, local_ni, local_i_start, local_no, local_o_start);
```

Comment:
Use pfft_ functions instead of pfftf_ so that it automagically works out the floating point precision

Reply:
I was under the impression (from the pfft user guide) that the pfft_ functions only accepted and returned double precision. The grids in meraxes/21CMFAST were single precision using fftwf so I just switched to pfftf. If I change the functions over to pfft_ what kind of casting will I need to do to the inputs/outputs of the functions?

Reply:
Ok, then I am confused about pfft, I think. What I thought happened was that the pfft.h file is compiled with a prototype which contains either float or double types depending on the "--enable-single" flag, but I guess not.

Note though that the version of pfft we compile doesn't support single precision floating point because the "--enable-single" flag is not passed in (which is why the travis build fails), so to keep using single precision grids you will need to alter the build system to turn that on.

Reply:
If there will be room in memory for these grids to be double precision (at the moment there are 6 grids using the fft) then I can just convert these to double precision.

Reply:
That's probably easiest for now - we can try it as a later optimization if this turns out to be a limitation.